src/cpp/core/system/PosixSystem.cpp

```diff
@@ -533,9 +539,89 @@ Error closeFileDescriptorsFrom(int fdStart)
 
    return Success();
 }
+#else
+// read the file descriptors from a virtual directory listing,
+// iteratore over them and close - much faster than the above method
+Error closeFileDescriptorsFrom(int fdStart)
+{
+   std::vector<unsigned int> fds;
+   Error error = getOpenFds(&fds);
+   if (error)
+      return error;
+
+   for (int fd : fds)
+   {
+      if (fd >= fdStart)
+      {
+         if (::close(fd) < 0 && errno != EBADF)
+            return systemError(errno, ERROR_LOCATION);
+      }
+   }
+
+   return Success();
+}
+#endif
 
 } // anonymous namespace
 
+Error getOpenFds(std::vector<unsigned int>* pFds)
+{
+   return getOpenFds(getpid(), pFds);
+}
+
+#ifndef __APPLE__
+Error getOpenFds(pid_t pid, std::vector<unsigned int>* pFds)
+{
+   std::string pidStr = safe_convert::numberToString(pid);
+   boost::format fmt("/proc/%1%/fd");
+   FilePath fdDir = FilePath(boost::str(fmt % pidStr));
+
+   std::vector<FilePath> children;
+   Error error = fdDir.children(&children);
+   if (error)
+      return error;
+
+   for (const FilePath& child : children)
+   {
+      try
+      {
+         unsigned int fd = boost::lexical_cast<unsigned int>(child.filename());
```

Comment:
Is the reason that `safe_convert` not used here that it isn't possible to supply a default value for the conversion that isn't also a possibly valid file descriptor? (not something you need to do for this PR but wondering if we should have some convert variants that use `boost::optional` or return a bool to make it possible to write these patterns w/o explicit exception handling)

Reply:
No urgency at all but I do think it's worth investing a couple of hours in. The pattern `safe_convert::thingTo(thing, ridiculousDefault) ... if (result == ridiculousDefault) { /* handle conversion failure */ }` (which we do use) isn't great even when you *do* have a usable `ridiculousDefault`.

Reply:
I will address this in a separate PR.